adam/learner/learner_utils.py

```diff
@@ -781,3 +790,31 @@ def get_slot_from_semantic_node(
         if object_node.concept == object_concept:
             return slot_var.name
     return slot
+
+
+def get_root_object_cluster_perception(
+    graph: DiGraph, matched_subgraph_nodes: AbstractSet[PerceptionGraphNode]
+) -> Optional[ObjectClusterNode]:
+    """Get the root object cluster node from a graph given a subset of nodes which represent the object.
```

Comment:
The function name and description are a bit confusing here, and I'd suggest changing them for clarity. It looks like this function is designed (as it should be) so that it doesn't crash when the subset of nodes is not actually a match for an object pattern match but say an action. So the name and docstring should reflect that.

(Re: the docstring, I think it's fine to leave this as a note below the single-line description since the description is already longish. Your call how to format it though.)

At the moment, the match should always be an object, but seems worth being clear that this should still work in the non-object case.

Reply:
For a name, maybe something like `get_root_if_perception_is_object_cluster`?